module/zfs/zfs_replay.c

```diff
@@ -473,8 +486,8 @@ zfs_replay_create(void *arg1, void *arg2, boolean_t byteswap)
 	xva.xva_vattr.va_nblocks = lr->lr_gen;
 	xva.xva_vattr.va_fsid = dnodesize;
 
-	error = dmu_object_info(zfsvfs->z_os, objid, NULL);
-	if (error != ENOENT)
+	error = check_foid_available(zfsvfs, lr->lr_foid);
```

Comment:
The `lr_foid` will have the dnodeslots encoded in the top 8 bits.  See the comment above ` LR_FOID_GET_SLOTS`.  We're going to need to use it to check all of the relevant slots, otherwise we still might get an `ENOSPC` error when performing the claim.  I'd suggest we simply do this and not bother with the helper function.

```diff
index 1443817..85d70ae 100644
--- a/module/zfs/zfs_replay.c
+++ b/module/zfs/zfs_replay.c
@@ -473,10 +473,14 @@ zfs_replay_create(void *arg1, void *arg2, boolean_t bytesw
        xva.xva_vattr.va_nblocks = lr->lr_gen;
        xva.xva_vattr.va_fsid = dnodesize;
 
-       error = dmu_object_info(zfsvfs->z_os, objid, NULL);
-       if (error != ENOENT)
+       dnode_t *dn;
+       error = dnode_hold_impl(zfsvfs->z_os, objid, DNODE_MUST_BE_FREE,
+           dnodesize >> DNODE_SHIFT, FTAG, &dn);
+       if (error)
                goto out;
 
+       dnode_rele(dn, FTAG);
+
        if (lr->lr_common.lrc_txtype & TX_CI)
                vflg |= FIGNORECASE;
```

Then let the bottom half of `zil_replay_log_record` handle this by forcing the txg sync and retrying.

```c
        ... 
        error = zr->zr_replay[txtype](zr->zr_arg, zr->zr_lr, zr->zr_byteswap);
        if (error != 0) {
                /*
                 * The DMU's dnode layer doesn't see removes until the txg
                 * commits, so a subsequent claim can spuriously fail with
                 * EEXIST. So if we receive any error we try syncing out
                 * any removes then retry the transaction.  Note that we
                 * specify B_FALSE for byteswap now, so we don't do it twice.
                 */
                txg_wait_synced(spa_get_dsl(zilog->zl_spa), 0);
                error = zr->zr_replay[txtype](zr->zr_arg, zr->zr_lr, B_FALSE);
                if (error != 0)
                        return (zil_replay_error(zilog, lr, error));
        }
```

Reply:
Oh, you're right. I missed that part because zfs_replay_create_acl doesn't have this...